### DISCUSSION TITLE
How to join tables using the Query Builder

### DISCUSSION THREAD

**Discussion by thiagomini** [mikro-orm/mikro-orm#4646] [Q&A] [answered]
I have defined the following schema for a Virtual Entity:

```ts
export interface CalendarSyncEvent {
  id: Id;
  meetingType: string;
  status: string;
  availabilityType: string;
}

export const calendarSyncEventSchema = new EntitySchema<CalendarSyncEvent>({
  name: 'CalendarSyncEventView',
  expression(em: EntityManager, where) {
    return em
      .createQueryBuilder<UserAuthRulesView>('CalendarSyncEventView')
      .select('*')
      .from('public.events')
      .join('offices', 'o', 'o.id = events.office_id')
      .where(where);
  },
  properties: {
    id: { type: Id },
    meetingType: { type: String },
    status: { type: String },
    availabilityType: { type: String },
    offices: {
      type: 'string[]',
    },
  },
});
```

However, when I try to execute the `repository.findAll()` method, it throws an error:

```
Metadata for entity public.events not found
```

Why is that throwing an error? I noticed it only throws when I add the `join()` clause. If I'm just using a standard `select`, it works fine. Should any metadata be necessary if we are using a query builder with explicit join clauses? 🤔

**Comment by B4nan** [comment, marked as answer]
And that code compiles? It shouldn't, the third param of `qb.join()` is not a string, it is an object. You need to think of the QB as a tool for working with the entities, it operates based on metadata. `qb.from()` expects entity name, and that is the main issue here most probably. Moreover, there is no reason to even use `qb.from()` since the entity type is already provided in the constructor. 

Same for the `offices`, that's wrong as well, the way you map it means it will be a scalar array property, yet you try to join it? You can only join relations.

Also, if you want to use virtual entities, you need to first define your entities (the regular ones) - and that is what you will be fetching via the `expression`. This looks like you are trying to create virtual entity that is fetching itself, maybe it works somehow, but its a misusage and undefined behaviour. Do you have an entity for the `events` table? As that is what you should be querying, that is what should be in the `createQueryBuilder` argument, not your virtual entity name.

**Reply by thiagomini** [reply]
I actually changed to `getKnex()` to use the query builer and join with tables not yet defined by our schemas. The intention was to make that virtual entity as a sort of a database view - to it actually joins many other tables that are not yet defined.
